raiden/network/proxies/secret_registry.py

```diff
@@ -82,10 +82,17 @@ def register_secret_batch(self, secrets: List[typing.Secret]):
             log.debug('registerSecretBatch skipped', **log_details)
             return
 
-        log.debug('registerSecretBatch called', **log_details)
+        gas_limit = self.proxy.estimate_gas('registerSecretBatch', secrets)
+        if not gas_limit:
+            log.critical(
+                'Call to registerSecretBatch will fail. Out of funds?',
+                **log_details,
+            )
+            raise RaidenUnrecoverableError('Call to registerSecretBatch will fail. Out of funds?')
```

Comment:
Hummm, out of funds would make sense, I missed it while reading the PR description.

My proposal then would be the following:

- Finish `Ensure the node works on a consistent view of the blockchain` #2885 , this will give less room for race conditions to happen, allowing us to "make sure" that `eth_estimateGas` and the pre conditions checks are working on a consistent view of the blockchain. (Questions here is which block should we use, a confirmed block or latest?)
- Always do the preconditions checks **before** the gas estimation, **iff** the gas estimation returns `None` raise an UnrecoverableError
  - Under this circumstance the account's gas must be checked. For functions which have a constant gas usage, this can be precisely done, for functions that have a variable gas usage (e.g. unlock), we can do a best effort scenario. This will provide a better message for debugging/the user.
  - Otherwise an "unexpected precondition was not met" error should be used, because that is a programming error which we don't know the side effects of.
- *Maybe* use a confirmed block for the fix for #2885 **and** do the [check for the `gas_reserve`](https://github.com/raiden-network/raiden/blob/f6f33e179d873b38b674d9db7d6baa5b44aff276/raiden/tasks.py#L68), if the given confirmed block doesn't have enough reserve the node will quit, meaning that  `None` from `estimateGas` was not because of lack of balance.

Reply:
> Finish Ensure the node works on a consistent view of the blockchain #2885 , this will give less room for race conditions to happen, allowing us to "make sure" that eth_estimateGas and the pre conditions checks are working on a consistent view of the blockchain. (Questions here is which block should we use, a confirmed block or latest?)

I started a bit differently.

I do a small part of #2885 here, only so that I can give the block_identifier to `estimate_gas()`. The block to check is the pending block. This is what geth does by default and geth is the only one who does not accept parametrization of estimate gas block. So we don't have much of a choice.

> Always do the preconditions checks before the gas estimation, iff the gas estimation returns None raise an UnrecoverableError 

 I am implementing something similar here (commits are not yet pushed). Let me finish it and you will look at it. Essentially the part of #2885 that I want to implement first here is the setting of the block number for the precondition checks. Then eth_estimateGas should work without race conditions. And then after that we can go for the full fledged solution on #2885 which I will start right after getting this PR merged.

Reply:
Why would you want to do the estimate gas before the preconditions checks? What does this achieve?

>  the question is can there be race conditions with the above strategy?

Yes, there will always be races, the preconditions checks are only best effort to avoid sending an unnecessary transactions, but races cannot prevented. And as you point out the races are handled properly, this is because of work for #1966.

> For open this should hold. No new block should invalidate preconditions. If a new block comes and someone else opened the channel that's already handled.

Using this as an example. The precondition "there is not channel among A-B before sending the transaction" can be invalidated by these events:

- User A and B wants to open channel at the same time
- Both run the precondition checks
- The transaction of one of the nodes will fail

The above shows when the precondition will be invalidated, the check prior to sending the transaction is to avoid an unnecessary transaction which wastes gas, the check after the transaction failed is to make debugging easier.

The problem is *not* about the preconditions being invalidated. The problem is having a **consistent** view of the blockchain.

I think your solution works, however it's confusing to me. We need to use a single block to make sure the data we retrieve for a given channel is consistent, for this reason I would stick to for all things in a proxy.

Note regarding the estimate gas call: I realized that it doesn't make sense to use the same block as the one for the checks, we should use latest here, so I would do this latest as to have the freshest view of the blockchain, perhaps this should be the pattern:

```python
preconditions()
gas = gas_estimate()
if gas is not None:
  txhash = send_transaction()
  result = poll(txhash)
if gas is None or result is Failure:
  detect_failure()
```

Reply:
> Note regarding the estimate gas call: I realized that it doesn't make sense to use the same block as the one for the checks (there is nothing to gain for it), we should use latest here.

Exactly. It does not make sense as no matter our blockchain view being consistent or not, estimate_gas only acts at the `pending` block at the moment. It gives us a preview of what would `send_transaction()` do right before we send it.

> Why would you want to do the estimate gas before the preconditions checks? What does this achieve?

I thought of it as an optimization and thought of precondition checks as only the blockchain checks before doing the call. If there are any other non-blockchain checks to do they should still happen. The optimization I thought is why do the blockchain checks before the estimateGas if once it fails we have to do them afterwards anyway to see if it failed for a known reasons or not. But taking the consistent view of the blockchain into account I got another idea to present in the next section below:


>  perhaps this should be the pattern:

I thought more on this and had a talk with @konradkonrad and came to a good understanding. It's a very similar pattern to what you suggest and to what I am implementing. Added a bit more details. You will notice I changed my mind on having the precondition checks only after the estimateGas. The comment explains why.

```python
nonblockchain_preconditions()
blockchain_preconditions(consistent_block)  # this will also function as a check of the logic that lead us to start the call in the first place
gas = gas_estimate(pending_block)
if gas is not None:
  txhash = send_transaction()
  result = poll(txhash)
if gas is None or result is Failure:
    transaction_mined = gas is not None
    known_race, msg = check_blockchain_conditions(mined_block if transaction_mined else pending_block)
    if known_race:
        raise RaidenRecoverableError('{} {}'.format('Transaction Failed' if transaction_mined else 'Transaction will fail', msg))
    else:
        raise RaidenUnrecoverableError("Sad face")  
```

What do you think on that?

Reply:
Yep, that is prett*y* good.

Just a side note, because of races the post conditions are not equal to the preconditions. An example would be channel deposit, which must have the channel open as a precondition, however if it races against a channel close, the post condition cannot assert on that.